feat: adding styles guidance (#6)
nodeshift/web-application-reference: docs/development/styles.md

# Styles

## Recommended Components

**TODO**

## Guidance

> **TODO**

Reviewer: mention:
need for out of box style system
or, make your own (if you have design team that may be working on a style system)

Reviewer: mention:
goal of the styles, who or what target device demographic is. mobile, desktop, browser types, browser versions,etc

### Frameworks

> * Plain CSS?

Reviewer: could split up to 3 categories: css, behavior, style system

css: sass, tailwind, less
style systems can be: tailwindui, bootstrap
behavior: headless ui, mui, radix

Reviewer: style systems can have existing design methodologies

Reviewer: javascript frameworks can have their own flavor of the above
styled-components, css-in-js, etc

Reviewer: sass nodejs library: https://github.com/sass/dart-sass, previous node-sass is deprecated

Reviewer: for plain css, could mention methodologies of organizing css/sass
```
https://getbem.com/introduction/
[OOCSS](http://oocss.org/) — Separating container and content with CSS “objects”
[SMACSS](http://smacss.com/) — Style-guide to write your CSS with five categories for CSS rules
[SUITCSS](https://suitcss.github.io/) — Structured class names and meaningful hyphens
[Atomic](https://acss.io/) — Breaking down styles into atomic, or indivisible, pieces
```

Reviewer 2: For design systems, we should definitely add Carbon, not just because it's an awesome style system but because its IBM's own creation, and we have a lot of experience with it.

Edit: oops, noticed that Carbon is already there.

* bootstrap
* tailwind

### Layout Methods

* grid/flex/other? - layout methods

* Media Queries?
  * here or in another section ?
  * maybe here with regard to cross platform?

### Preprocessors

A CSS preprocessor is a program that lets you generate CSS from the preprocessor's own unique syntax.

**Popular CSS preprocessors:**

* [Sass](https://sass-lang.com/)
* [LESS](https://lesscss.org/)

> * [PostCSS](https://postcss.org/)

Reviewer: i treat postcss different from preprocessor.
preprocess -> css -> postprocess with postcss

Reviewer: postcss plugins: https://github.com/postcss/autoprefixer

Another way of making an application responsive is with either the Grid or Flex layouts.  The team is familiar with both Grid and Flex, while Grid is slightly prefered.  Depending on the situation, both can be used.

Whichever layout method is used, Grid or Flex, will go hand in hand with using Media Queries

Of course, if you are using a CSS framework it is important to use whichever layout method is built in there.

### Preprocessors

While preprocessors are still being used, built-in css properties and rules are now starting to deprecate the use of preprocessors.

However, the team does prefer the use of the [Sass](https://sass-lang.com/).

Many of the popular front-end CSS and javascript frameworks used to create application have these processors built in, so if you are using a framework, it is always recommended to use what that framework is using.

### General

#### Units of Measurement

The team thought that it was important to stick with one for a consistent UI and consistecy across the application.  Some helpful recommendations that came out of our discussion where:

* rem - preferred for accesibilty and for fonts

> * borders - these could be in **px**

Reviewer 3: why?

> * div tags - should be more percentage based

Reviewer 3: Disagree with this as a general rule, especially without any explanation/guidance

Author: After watching the video of the meeting, there wasn't really any explanation, so i think we should talk through this and the other comment in the next meeting

As mentioned earlier, if you are using a CSS framework, it is important to use what that framework is using.

Make sure to use the `viewport` metatags width/height, which will be helpful for the gloabl container and absoulte positiong.  Here is an example `viewport` for a mobile application

```html
<meta name="viewport" content="width=device-width, initial-scale=1" />
```

#### Vendor prefixes

> Browser vendors would add prefixes to non-standard/new css properties for developers to experiment with.  The team recommends the use of the [autoprefixer](https://www.npmjs.com/package/autoprefixer) library, which is an industry standard.  While this might important when supporting older code, it is recommended to look at if and why you might need a prefix when creating new applicaitons.

Reviewer 3: should mention this is a postcss plugin and requires postcss, and maybe talk about postcss itself

# Styles

## Recommended Components

N/A

## Guidance

> During our discussions it became clear that there are multiple ways of styling and that there is a lot of choice, but there was one common idea that kept re-occuring.  If you are using a CSS framework, be sure to use what that framework is using.

Reviewer 4: *re-occurring
minor nit about a spelling update...

### CSS Frameworks

CSS frameworks are a great choice for developing applicaitons.  Some advantages are:

* Great for out-of-the-box prototyping
* customizable
* consistent

> [Tailwind CSS](https://tailwindcss.com/) is a modern choice that the team has used.  Other recommendations that the team made are [Carbon](https://carbondesignsystem.com/) and [Patternfly](https://www.patternfly.org/) which are both IBM and Red Hat projects respectivly.

Reviewer 4: *respectively
minor nit about a spelling update...

> It is also important that if you are using a front-end JavaScipt framework, the CSS framework should fit into it.

Reviewer 4: *JavaScript

minor nit about a spelling update...

Another way of making an application responsive is with either the Grid or Flex layouts.  The team is familiar with both Grid and Flex, while Grid is slightly prefered.  Depending on the situation, both can be used.

Whichever layout method is used, Grid or Flex, will go hand in hand with using Media Queries

Of course, if you are using a CSS framework it is important to use whichever layout method is built in there.

### Preprocessors

While preprocessors are still being used, built-in css properties and rules are now starting to deprecate the use of preprocessors.

However, the team does prefer the use of the [Sass](https://sass-lang.com/).

Many of the popular front-end CSS and javascript frameworks used to create application have these processors built in, so if you are using a framework, it is always recommended to use what that framework is using.

### General

#### Units of Measurement

The team thought that it was important to stick with one for a consistent UI and consistecy across the application.  Some helpful recommendations that came out of our discussion where:

* rem - preferred for accesibilty and for fonts

> * borders - these could be in **px** - decrotavie / doesn't scale well

Reviewer 4: *decorative
minor nit about a spelling update...

  * Borders are mostly decorative anyway, and using doesn't scale well when using something that are not pixels

As mentioned earlier, if you are using a CSS framework, it is important to use what that framework is using.

> Make sure to use the `viewport` metatags width/height, which will be helpful for the gloabl container and absoulte positiong.  Here is an example `viewport` for a mobile application

Reviewer 4: *global and *absolute

minor nit about some spelling updates that are needed...

CSS frameworks are a great choice for developing applications.  Some advantages are:

* Great for out-of-the-box prototyping
* customizable
* consistent

[Tailwind CSS](https://tailwindcss.com/) is a modern choice that the team has used.  Other recommendations that the team made are [Carbon](https://carbondesignsystem.com/) and [Patternfly](https://www.patternfly.org/) which are both IBM and Red Hat projects respectively.

It is also important that if you are using a front-end JavaScipt framework, the CSS framework should fit into it.

### Layout Methods

Since it is generally harder to design for a smaller screen, the team recommends a mobile first and responsive approach to an applications layout.  When defining css rules it is recommended to put the rules for mobile first followed by the rules for other layouts that your application might have.

The use of [Media Queries](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_media_queries/Using_media_queries) allows you to apply CSS styles to your application using various characteristics, such as screen resolution and orientation.  Media Queries will help with the responsive approach to writing applications as mentioned above.

> Another way of making an application responsive is with either the Grid or Flex layouts.  The team is familiar with both Grid and Flex, while Grid is slightly prefered.  Depending on the situation, both can be used.

Reviewer 3: ```suggestion
Another way of making an application responsive is with either the [Grid](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_grid_layout) or [Flexbox](https://developer.mozilla.org/en-US/docs/Learn/CSS/CSS_layout/Flexbox) layout methods.  The team is familiar with both Grid and Flexbox, while Grid is slightly preferred.  Depending on the situation, both can be used.

The most significant distinction between Grid and Flexbox, is that Grid is a two-dimensional layout method and Flexbox is a one-dimensional layout method.
```

> Another way of making an application responsive is with either the Grid or Flex layouts.  The team is familiar with both Grid and Flex, while Grid is slightly prefered.  Depending on the situation, both can be used.

Reviewer 3: ```suggestion
Another way of making an application responsive is with either the [Grid](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_grid_layout) or [Flexbox](https://developer.mozilla.org/en-US/docs/Learn/CSS/CSS_layout/Flexbox) layout methods.  The team is familiar with both Grid and Flexbox, while Grid is slightly preferred.  Depending on the situation, both can be used.

The most significant distinction between Grid and Flexbox, is that Grid is a two-dimensional layout method and Flexbox is a one-dimensional layout method.
```

Reviewer 3: I'm fine if we want to say Grid is slightly preferred, but if we are going to say that, we should explain why.  Also, my opinion is that if your requirement is for a one-dimensional layout, then Flexbox is preferable because there's less to fiddle with.

Whichever layout method is used, Grid or Flex, will go hand in hand with using Media Queries

Of course, if you are using a CSS framework it is important to use whichever layout method is built in there.

### Preprocessors

While preprocessors are still being used, built-in css properties and rules are now starting to deprecate the use of preprocessors.  It is also important to decide if the scope and scale of your project needs one.

However, the team does prefer the use of the [Sass](https://sass-lang.com/).

> The team also agreed that it was important to be more CSS centric(native CSS features) even when using a preprocessor.  This was because the language is introducing more and more features, which will start to make these preprocessors deprecated.

Reviewer 3: ```suggestion
The team also agreed that it was important to maximize use of native CSS features, even when using a preprocessor.  This is because the language is introducing more and more features, which will start to make these preprocessors deprecated.
```
